keps/sig-network/4770-endpointslice-controller-flexibility/README.md

```diff
@@ -0,0 +1,481 @@
+
+The `SetLabelsAnnotations` function implements `SetEndpointSliceLabelsAnnotations`. The Endpoints (EndpointSlice Mirroring Controller) strategy is to set the "kubernetes.io/service-name" and "endpointslice.kubernetes.io/managed-by" labels, clones the labels and annotations from the Endpoints and updates the EndpointSlice if there is any changes.
+
+```golang
+// DesiredEndpointSlicesFromServicePods returns the list of desired endpointslices for the given endpoints.
+// It also return which address types are handled.
+func DesiredEndpointSlicesFromEndpoints(
+	endpoints *corev1.Endpoints,
+	maxEndpointsPerSubset int32,
+) ([]*EndpointPortAddressType, sets.Set[discovery.AddressType])
+
+type LabelsAnnotationsFromEndpoints struct {
+	Endpoints *corev1.Endpoints
+}
+
+func (lafe *LabelsAnnotationsFromEndpoints) SetLabelsAnnotations(logger klog.Logger, epSlice *discovery.EndpointSlice, controllerName string) (map[string]string, map[string]string, bool) 
+```
+
+Example of usage replacing the current `Reconcile` call from the EndpointSlice Mirroring Controller ([pkg/controller/endpointslicemirroring/endpointslicemirroring_controller.go#L334](https://github.com/kubernetes/kubernetes/blob/v1.30.2/pkg/controller/endpointslicemirroring/endpointslicemirroring_controller.go#L334)):
+```golang
+desiredEndpoints, supportedAddressesTypes := endpointslicerec.DesiredEndpointSlicesFromEndpoints(endpoints, r.maxEndpointsPerSubset)
+return rec.Reconcile(
+    logger,
+    endpoints,
+    endpoints,
+    desiredEndpoints,
+    existingSlices,
+    supportedAddressesTypes,
+    nil, // EndpointSlice Mirroring doesn't use traffic distribution.
+    labelsAnnotationsFromEndpoints.SetLabelsAnnotations,
+    time.Time{}, // So endpoints.kubernetes.io/last-change-trigger-time will be removed.
+)
+```
+
+#### Metrics
+
+Most of the metrics exposed by the EndpointSlice and EndpointSlice Mirroring Controllers are similar. The EndpointSlice Controller exposes `endpointslices_changed_per_sync`, `syncs` and `services_count_by_traffic_distribution` that are not exposed by the EndpointSlice Mirroring Controller. The EndpointSlice Mirroring Controller exposes `endpoints_updated_per_sync`, `addresses_skipped_per_sync` and `endpoints_sync_duration` that are not exposed by the EndpointSlice Controller. Metrics exposed will need to be merged together, and the metrics package should be changed to support any type of resources (pods/service, endpoints...).
+
+### Test Plan
+
+[x] I/we understand the owners of the involved components may require updates to
+existing tests to make this code solid enough prior to committing the changes necessary
+to implement this enhancement.
+
+##### Prerequisite testing updates
+
+##### Unit tests
+
+* Tests in `pkg/controller/endpointslicemirroring/reconciler_test.go` will be moved to `staging/src/k8s.io/endpointslice`
+
+##### Integration tests
+
+TDB
+
+##### e2e tests
+
+- Usage of `service.kubernetes.io/endpoint-controller-name` on services
```

Comment:
these tests are better done in integration, there is no e2e functionality here , is only API operations

Reply:
Similar test for the `service.kubernetes.io/service-proxy-name` is an e2e test (https://github.com/kubernetes/kubernetes/blob/v1.30.3/test/e2e/network/service.go#L2246). Do you consider it different? Does this really has to be moved to integration test?

Reply:
that test linked exercises the dataplane, if you were expecting to test also the dataplane then I stand corrected, but I'd like to see the copy of the api only tests in the integration too :)

Reply:
I understand now, thank you for the explanation, I moved it.